### PR TITLE
Add support for heterogeneous reactions in MSContactor

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
Whilst developing a model fora leaching operating for PrOMMiS, it was realized that there was a need for the `MSContactor` model to support reactions which involve multiple streams (i.e. heterogeneous reactions).

## Changes proposed in this PR:
- Add configuration arguments to `MSContactor` to add heterogeneous reactions
- Add necessary variables and constraints to `MSContactor`
- Unit testing and documentation

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by eslickj — APPROVED** [review]

File: docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst [hunk 1]
```diff
@@ -57,41 +60,43 @@ Variables
...
 
-.. math:: 0 = \sum_p{F_{t,x-,s,p,j}} - \sum_p{F_{t,x,s,p,j}} + \left[ \sum_p{F_{side,t,x,s,p,j}} \right] + \sum_o{M_{t,x,s,o,j}} + \left[ \sum_p{G_{rate,t,x,s,p,j}} + \sum_p{G_{equil,t,x,s,p,j}} + \sum_p{G_{inher,t,x,s,p,j}} \right]
+.. math:: 0 = \sum_p{F_{t,x-,s,p,j}} - \sum_p{F_{t,x,s,p,j}} + \left[ \sum_p{F_{side,t,x,s,p,j}} \right] + \sum_o{M_{t,x,s,o,j}} + \left[ \sum_p{G_{rate,t,x,s,p,j}} + \sum_p{G_{equil,t,x,s,p,j}} + \sum_p{G_{inher,t,x,s,p,j}} + \sum_p{G_{hetero,t,x,s,p,j}} \right]
 
 where ``F`` is the material flow term, ``x-`` represents the the previous finite element (``x-1`` in the case of co-current flow and ``x+1`` in the case of counter-current flow), ``F_side`` is the material flow term for a side stream (if present) and ``o`` represents all other streams in the model (for cases where ``s`` is the second index (i.e., M_{t,x,o,s,j}) the term is multiplied by -1). The reaction generation terms are only included if the appropriate reaction type is supported by the reaction or property package for the stream.
```
Comment (agarciadiego) [inline docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst:87]: "the the"

File: docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst [hunk 2]
```diff
@@ -116,21 +121,150 @@ MSContactor Class
...
+            add_object_reference(self, "_params", self.config.parameters)
+
+            # Need to define a reaction rate variable - make sure units are correct
+            self.reaction_rate = Var(
+                self.params.reaction_idx,
+                initialize=0,
+                units=#units
+                )
+
+            # Define rule for calcuating reaction rate
```
Comment (agarciadiego) [inline docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst:261]: calculating

**Review by agarciadiego — APPROVED** [review]
two small typos

File: docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst (hunk 1 above)
Comment (agarciadiego) [reply docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst:87]: first line of 87

**Pull request merged**